### ISSUE TITLE
mdPanel: Scope Watchers not Cleaned Up on Close

### ISSUE THREAD

**Issue by chrisguerrero** [angular/material#8997]
**Actual Behavior**:
 -  `What is the issue? *` After an `MdPanelRef` reference is stored, and the `close()` and `open()` methods are subsequently called, the panel's scope watchers aren't cleaned up. This causes a memory leak until the `destroy()` method is manually called on the panel reference.
 -  `What is the expected behavior?` A panel's scope watchers should be deregistered when the `close()` method is called.

**CodePen** (or steps to reproduce the issue): *
 -  `CodePen Demo which shows your issue:` https://codepen.io/chrisguerrero/pen/xOdAxr
 -  `Details:` Open and close the dialog repeatedly and inspect the watcher count from the panel's scope and $rootScope. Three scope watchers are added every time the panel is opened.

If you click the Destroy button, the watchers are deregistered as expected. The next time the panel is opened, the `create()` method is called again and the process repeats itself with a new panel scope.

**Angular Versions**: *
 -  `Angular Version:` 1.5.5
 -  `Angular Material Version:` 1.1.0-rc.5-master-d593229 (latest release)

**Additional Information**:
 -  `Browser Type: *` all
 -  `Browser Version: *` all
 -  `OS: *` all
 -  `Stack Traces:` none

----
Shortcut to create a [new CodePen Demo](http://codepen.io/team/AngularMaterial/pen/bEGJdd).
Note: `*` indicates required information. Without this information, your issue may be auto-closed.

> Do not modify the titles or questions. Simply add your responses to the ends of the questions.
  Add more lines if needed.

**Comment by crisbeto** [comment]
I believe that this is on purpose. The point of closing it is to be able to re-open and re-use the same instance later on.

**Comment by chrisguerrero** [comment]
If that's the case, shouldn't the watchers not be re-added when the panel is opened again? 

In my implementation, I have a complex panel template with a form and about 400 scope listeners. The form is intended to be accessed from anywhere in the app, so I store the `MdPanelRef` to avoid having to re-create a new panel every time.  Currently, the number of watchers in my app jumps by 400 each time the panel is shown, quickly bringing my app to a grinding halt during the digest loop.

Even if the watchers weren't re-added when the panel opens, does it make sense to have those 400 watchers being checked on each digest loop when the panel is closed? Imagine there are 10 panel references stored, each with 400 scope listeners.

**Comment by crisbeto** [comment]
If you really don't want those 400 watchers staying in the background, you can call `destroy` after `close`. It would be complicated/error prone if Material tried to collect and re-add all the watchers every time a panel was opened/closed.

**Comment by chrisguerrero** [comment]
The issue still remains that additional watchers are added each time `open()` is called.

**Comment by crisbeto** [comment]
There was actually an issue where `destroy` wasn't destroying the scope either. This is fixed that was merged into master https://github.com/angular/material/pull/8848

**Comment by chrisguerrero** [comment]
That commit is included in the version of angular-material in my codepen (1.1.0-rc.5-master-d593229) so I think that's unrelated.

**Comment by crisbeto** [comment]
@chrisguerrero I got a chance to look at this again. From what I'm seeing it works fine. If you change the `closeDialog` method to this, the watchers will get cleaned up after the close. Of course, any subsequent panels won't be compiled since you need to create a new panel instance:

```js
PanelDialogCtrl.prototype.closeDialog = function() {
  var panelRef = this._mdPanelRef;
  return panelRef.close().then(function() {
    angular.element(document.querySelector('.demo-dialog-open-button')).focus();
    panelRef.destroy();
  });
};
```

**Comment by chrisguerrero** [comment]
@crisbeto Yeah, that's what I ended up doing as a workaround. I actually have `panelRef.destroy()` inside the `onDomRemoved` callback for the panel, since the user can press the escape key and close the panel without calling the `closeDialog()` method.

It seems to me that intuitively, if you have a panel that's intended to be re-used, you should only create it once, and then open/close a single panel reference. Only once you know the panel should never be opened again should you destroy it. Currently, you must create, destroy, and create a new panel reference each time.

**Comment by crisbeto** [comment]
The panel is meant to be fairly low-level @chrisguerrero, in order to be able to accommodate as many use-cases as possible, hence why you need to do a little extra work when you're using it. I'm closing this for now since it seems like the panel is working as expected.

**Comment by vas1468** [comment]
I agree with @chrisguerrero and this issue needs more investigation..I ended up doing what suggested above ( `panelRef.destroy()` inside the `onDomRemoved` callback)
